halo2_proofs/examples/poly-eval.rs

```diff
@@ -0,0 +1,310 @@
+use std::marker::PhantomData;
+
+use halo2_proofs::{
+    arithmetic::Field,
+    circuit::{AssignedCell, Chip, Layouter, Region, SimpleFloorPlanner, Value},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error, Instance, Fixed, Selector},
+    poly::Rotation,
+};
+
+const EVALADVCOL:usize = 4;
+
+trait PolyEvalInstructions<F: Field>: Chip<F> {
+    /// Variable representing a coefficient.
+    /// Certainly it could be from another Field/Group,
+    /// orther than current one
+    type Coeff;
+
+    /// Field element
+    type Elem;
+
+    /// Load Element to be used as eval point 
+    fn load_element(&self, layouter: impl Layouter<F>, a: Value<F>) -> Result<Self::Elem, Error>;
+
+    /// Evaluation
+    fn eval(&self, layouter: impl Layouter<F>, coeffs: Vec<Self::Coeff>, elem: Self::Elem) -> Result<Self::Elem, Error>;
+
+    /// Expose the calculation results as a target whom instance compares with
+    fn reveal(&self, layouter: impl Layouter<F>, elem: Self::Elem, row: usize) -> Result<(), Error>;
+    
+}
+
+/// Chip for poly eval
+#[derive(Clone, Debug)]
+struct PolyEvalConfig {
+    /// cols: | coeff | point | result | accumulative result |
+    advice: [Column<Advice>; EVALADVCOL],
+
+    /// This is the public input (instance) column
+    instance: Column<Instance>,
+
+    /// | X as kth unit of root |
+    point: Column<Fixed>,
+
+    /// A selector controls gate
+    e_sel: Selector,
+}
+
+struct PolyEvalChip<F: Field> {
+    config: PolyEvalConfig,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> PolyEvalChip<F> {
+    fn construct(config: <Self as Chip<F>>::Config) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+
+    fn configure(
+        meta: &mut ConstraintSystem<F>,
+        advice: [Column<Advice>; EVALADVCOL],
+        instance: Column<Instance>,
+        point: Column<Fixed>,
+    ) -> <Self as Chip<F>>::Config {
+        for col in advice.iter() {
+            meta.enable_equality(*col);
+        }
+        meta.enable_equality(instance);
+        meta.enable_constant(point);
+        // All constraints enforced at once
+        let e_sel = meta.selector();
+
+        meta.create_gate("x pow", |meta| {
+            let cur_x = meta.query_advice(advice[1], Rotation::cur());
+            let next_x = meta.query_advice(advice[1], Rotation::next());
+            let x = meta.query_fixed(point, Rotation::cur());
+            let pow_s = meta.query_selector(e_sel);
+
+            vec![pow_s * (next_x - cur_x * x)]
+        });
```

Comment:
This is not checking that the first `acc` is indeed equal to `0`.

This probably needs a new selector `is_first` or similar.

Reply:
You are right. Will add it.